MacPass/MPPasswordInputController.m

```diff
@@ -138,6 +147,150 @@ - (IBAction)_submit:(id)sender {
+  if (status != errSecSuccess) {
+    [self _createAndAddRSAKeyPair];
+    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)getquery, (CFTypeRef *)&publicKey);
+    if (status != errSecSuccess) {
+      NSString* description = (__bridge NSString*)SecCopyErrorMessageString(status, NULL);
+      NSLog(@"Error while trying to query public key from Keychain: %@", description);
+      return;
+    }
+  }
+  SecKeyAlgorithm algorithm = kSecKeyAlgorithmRSAEncryptionOAEPSHA512;
+  BOOL canEncrypt = SecKeyIsAlgorithmSupported(publicKey, kSecKeyOperationTypeEncrypt, algorithm);
+  if(canEncrypt) {
+    int k = (int)SecKeyGetBlockSize(publicKey);
+    int hlen = 512 / 8;
+    int maxMessageLengthInByte = k - 2 * hlen - 2;
+    if([passwordData length] <= maxMessageLengthInByte) {
+      CFErrorRef error = NULL;
+      NSData* cipherText = (NSData*)CFBridgingRelease(SecKeyCreateEncryptedData(publicKey, algorithm, (__bridge CFDataRef)passwordData, &error));
+      if (cipherText) {
+        [touchIDSecuredPasswords setObject:cipherText forKey:databaseId];
+      }
+      else {
+        NSError *err = CFBridgingRelease(error);
+        NSLog(@"Error while trying decrypt password for TouchID unlock: %@", [err description]);
+      }
+    }
+    else {
+      NSLog(@"The password is to large to be encrypted");
+      return;
+    }
+  }
+  else {
+      NSLog(@"The key retreived from the Keychain is unable to encrypt data");
+  }
+  if (publicKey)  { CFRelease(publicKey);  }
+}
+
+- (NSString*) _loadPasswordForTochIDUnlock: (NSString*) databaseId {
+  NSString* result = nil;
+  NSData* cipherText = [touchIDSecuredPasswords valueForKey:databaseId];
+  if(cipherText != nil) {
+    NSData* tag = [@"com.hicknhacksoftware.macpass.privatekey" dataUsingEncoding:NSUTF8StringEncoding];
+    NSDictionary *queryPrivateKey = @{
+      (id)kSecClass: (id)kSecClassKey,
+      (id)kSecAttrApplicationTag: tag,
+      (id)kSecAttrKeyType: (id)kSecAttrKeyTypeRSA,
+      (id)kSecReturnRef: @YES,
+    };
+    SecKeyRef privateKey = NULL;
+    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)queryPrivateKey, (CFTypeRef *)&privateKey);
+    if (status == errSecSuccess) {
+      SecKeyAlgorithm algorithm = kSecKeyAlgorithmRSAEncryptionOAEPSHA512;
+      BOOL canDecrypt = SecKeyIsAlgorithmSupported(privateKey, kSecKeyOperationTypeDecrypt, algorithm);
+      if(canDecrypt) {
+        if([cipherText length] == SecKeyGetBlockSize(privateKey)) {
+          CFErrorRef error = NULL;
+          NSData* clearText = (NSData*)CFBridgingRelease(SecKeyCreateDecryptedData(privateKey, algorithm, (__bridge CFDataRef)cipherText, &error));
+          if (clearText) {
+            result = [[NSString alloc]initWithData:clearText encoding:NSUTF8StringEncoding];
+          }
+          else {
+            NSError *err = CFBridgingRelease(error);
+            NSLog(@"Error while trying decrypt password for TouchID unlock: %@", [err description]);
```

Comment:
Missing `to` between `trying` and `decrypt`.

Reply:
Fixed. Thanks for the review.